File: org_runtime/drift.py
# file: org_runtime/drift.py
"""
Drift Comparator — pure function, no side effects.

Computes a structured diff between two state dicts.
No Engine dependency. Structural density is calculated
directly from the dict using the same formula as graph.py:

    density = len(dependencies) / (n * (n - 1))   if n >= 2
            = 0.0                                  otherwise
"""

from __future__ import annotations

from typing import Any, Dict, List, Set


def compare_states(state_a: dict, state_b: dict) -> dict:
    """
    Compare two state dicts and return a structured diff.

    Both inputs are expected to be the output of OrgState.to_dict().

    Returns dict with:
        role_count_delta, active_role_delta, structural_debt_delta,
        structural_density_delta, added_roles, removed_roles,
        activated_roles, deactivated_roles
    """
    roles_a = state_a.get("roles", {})
    roles_b = state_b.get("roles", {})

    ids_a: Set[str] = set(roles_a.keys())
    ids_b: Set[str] = set(roles_b.keys())

    added = sorted(ids_b - ids_a)
    removed = sorted(ids_a - ids_b)

    # Active role counts
    active_a = sum(1 for r in roles_a.values() if r.get("active", True))
    active_b = sum(1 for r in roles_b.values() if r.get("active", True))

    # Activation changes (roles present in both states)
    common = ids_a & ids_b
    activated: List[str] = []
    deactivated: List[str] = []
    for rid in sorted(common):
        was_active = roles_a[rid].get("active", True)
        is_active = roles_b[rid].get("active", True)
        if not was_active and is_active:
            activated.append(rid)
        elif was_active and not is_active:
            deactivated.append(rid)

    # Structural density — computed directly, no Engine
    density_a = _compute_density(state_a)
    density_b = _compute_density(state_b)

    debt_a = state_a.get("structural_debt", 0)
    debt_b = state_b.get("structural_debt", 0)

    return {
        "role_count_a": len(roles_a),
        "role_count_b": len(roles_b),
        "role_count_delta": len(roles_b) - len(roles_a),
        "active_role_a": active_a,
        "active_role_b": active_b,
        "active_role_delta": active_b - active_a,
        "structural_debt_a": debt_a,
        "structural_debt_b": debt_b,
        "structural_debt_delta": debt_b - debt_a,
        "structural_density_a": round(density_a, 4),
        "structural_density_b": round(density_b, 4),
        "structural_density_delta": round(density_b - density_a, 4),
        "added_roles": added,
        "removed_roles": removed,
        "activated_roles": activated,
        "deactivated_roles": deactivated,
    }


def _compute_density(state_dict: dict) -> float:
    """
    Pure density calculation from a state dict.
    Same formula as graph.compute_structural_density:
        edges / (n * (n - 1))   for directed graph, n >= 2
    """
    roles = state_dict.get("roles", {})
    deps = state_dict.get("dependencies", [])
    n = len(roles)
    if n < 2:
        return 0.0
    max_edges = n * (n - 1)
    return len(deps) / max_edges

File: org_runtime_rust/src/drift.rs
//! Drift detection — determinism verification and state comparison.
//!
//! All numeric values are fixed-point i64 (SCALE = 10_000).
//! No float arithmetic anywhere.

use std::collections::BTreeSet;

use org_engine_replica::domain::OrgState;
use org_engine_replica::events::EventEnvelope;
use org_engine_replica::graph::compute_structural_density;

use crate::replay;

/// Verify determinism by replaying the same events twice and
/// asserting identical hashes. Panics on failure.
pub fn verify_determinism(events: &[EventEnvelope]) {
    let (_, hash1) = replay::rebuild_state(events);
    let (_, hash2) = replay::rebuild_state(events);

    if hash1 != hash2 {
        panic!(
            "DETERMINISM FAILURE: two replays produced different hashes.\n\
             Run 1: {}\n\
             Run 2: {}",
            hash1, hash2
        );
    }
}

/// Structured state comparison — all values are integers.
///
/// Returns a DriftReport with deltas and role lifecycle changes.
pub fn compare_states(state_a: &OrgState, state_b: &OrgState) -> DriftReport {
    let ids_a: BTreeSet<&str> = state_a.roles.keys().map(|s| s.as_str()).collect();
    let ids_b: BTreeSet<&str> = state_b.roles.keys().map(|s| s.as_str()).collect();

    let added: Vec<String> = ids_b
        .difference(&ids_a)
        .map(|s| s.to_string())
        .collect();
    let removed: Vec<String> = ids_a
        .difference(&ids_b)
        .map(|s| s.to_string())
        .collect();

    let active_a = state_a.roles.values().filter(|r| r.active).count() as i64;
    let active_b = state_b.roles.values().filter(|r| r.active).count() as i64;

    // Activation changes in roles present in both states
    let common = ids_a.intersection(&ids_b);
    let mut activated = Vec::new();
    let mut deactivated = Vec::new();
    for rid in common {
        let was_active = state_a.roles[*rid].active;
        let is_active = state_b.roles[*rid].active;
        if !was_active && is_active {
            activated.push(rid.to_string());
        } else if was_active && !is_active {
            deactivated.push(rid.to_string());
        }
    }

    // Structural density (int64 fixed-point, no float)
    let density_a = compute_structural_density(state_a);
    let density_b = compute_structural_density(state_b);

    DriftReport {
        role_count_a: state_a.roles.len() as i64,
        role_count_b: state_b.roles.len() as i64,
        role_count_delta: state_b.roles.len() as i64 - state_a.roles.len() as i64,
        active_role_a: active_a,
        active_role_b: active_b,
        active_role_delta: active_b - active_a,
        structural_debt_a: state_a.structural_debt,
        structural_debt_b: state_b.structural_debt,
        structural_debt_delta: state_b.structural_debt - state_a.structural_debt,
        structural_density_a: density_a,
        structural_density_b: density_b,
        structural_density_delta: density_b - density_a,
        added_roles: added,
        removed_roles: removed,
        activated_roles: activated,
        deactivated_roles: deactivated,
    }
}

/// Structured drift report — all numeric fields are i64.
#[derive(Debug, Clone)]
pub struct DriftReport {
    pub role_count_a: i64,
    pub role_count_b: i64,
    pub role_count_delta: i64,
    pub active_role_a: i64,
    pub active_role_b: i64,
    pub active_role_delta: i64,
    pub structural_debt_a: i64,
    pub structural_debt_b: i64,
    pub structural_debt_delta: i64,
    pub structural_density_a: i64,
    pub structural_density_b: i64,
    pub structural_density_delta: i64,
    pub added_roles: Vec<String>,
    pub removed_roles: Vec<String>,
    pub activated_roles: Vec<String>,
    pub deactivated_roles: Vec<String>,
}
